### PR TITLE
8328638: Fallback option for POST-only OCSP requests

### PR DESCRIPTION
**Pull request by shipilev**
See the rationale/discussion in the bug. This patch introduces the option that allows to restore pre-[JDK-8179503](https://bugs.openjdk.org/browse/JDK-8179503) behavior. The default behavior does not change. Better suggestions for flag name are welcome.

Additional testing:
 - [x] `jdk_security` passes out of the box (includes new test config)
 - [x]  `jdk_security` passes with flag override
 - [x] Eyeballing `GetPostTests` amended debugging output, `GET`-s are used by default for small requests, `POST`-s are used for everything with flag override

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8328810](https://bugs.openjdk.org/browse/JDK-8328810) to be approved

### Issues
 * [JDK-8328638](https://bugs.openjdk.org/browse/JDK-8328638): Fallback option for POST-only OCSP requests (**Enhancement** - P4)
 * [JDK-8328810](https://bugs.openjdk.org/browse/JDK-8328810): Fallback option for POST-only OCSP requests (**CSR**)

### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)
 * [Rajan Halade](https://openjdk.org/census#rhalade) (@rhalade - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18408/head:pull/18408` \
`$ git checkout pull/18408`

Update a local copy of the PR: \
`$ git checkout pull/18408` \
`$ git pull https://git.openjdk.org/jdk.git pull/18408/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18408`

View PR using the GUI difftool: \
`$ git pr show -t 18408`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18408.diff">https://git.openjdk.org/jdk/pull/18408.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18408#issuecomment-2010520871)

### CODE REVIEW

**Comment by seanjmullan** [comment]
Thanks for doing this - I think it is a fine idea to have a fallback option to use GET. It does need a CSR though since you are introducing a new system property.

**Comment by seanjmullan** [comment]
Ideally, we should also modify the tests in `test/jdk/security/infra/java/security/cert/CertPathValidator/certification` to test OCSP with POST. I think it should be easy enough to add an additional line to each test like:

` * @run main/othervm/timeout=180 -Djava.security.debug=certpath -Dcom.sun.security.ocsp.useget=false ActalisCA OCSP`
 
 This means we would be testing real OCSP responders supported by CAs in the Java Root Program, and not just a test responder that we created.
 
 @rhalade would this change be ok with you?

**Comment by shipilev** [comment]
> Thanks for doing this - I think it is a fine idea to have a fallback option to use POST. It does need a CSR though since you are introducing a new system property.

All right, good! How do you feel about the option name? Would like to arrive to consensus for it before I plaster it all over CSR text and tests.

**Comment by seanjmullan** [comment]
> > Thanks for doing this - I think it is a fine idea to have a fallback option to use POST. It does need a CSR though since you are introducing a new system property.
> 
> All right, good! How do you feel about the option name? Would like to arrive to consensus for it before I plaster it all over CSR text and tests.

Yes, I am fine with the name.

**Comment by shipilev** [comment]
I amended the `CAInterop` test (unfortunately, it fails even without the patch, I think due to [JDK-8324583](https://bugs.openjdk.org/browse/JDK-8324583)), and filed the CSR.

**Review by seanjmullan — APPROVED** [review]
Looks good. Please also write a release note and I will review it.

**Comment by shipilev** [comment]
> Looks good. Please also write a release note and I will review it.

Here you go: https://bugs.openjdk.org/browse/JDK-8329111

File: test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java [hunk 1]
```diff
@@ -433,6 +476,7 @@
  * @library /test/lib
  * @build jtreg.SkippedException ValidatePathWithURL CAInterop
  * @run main/othervm -Djava.security.debug=certpath,ocsp CAInterop emsignrootcag1 OCSP
```
Comment (rhalade) [inline test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java:478]: eMudhra OCSP responders don't support GET calls so these tests fail, refer - [JDK-8328830](https://bugs.openjdk.org/browse/JDK-8328830). We are following up with CA. Meanwhile, can you please add these test to ProblemList at https://github.com/openjdk/jdk/blob/master/test/jdk/ProblemList.txt#L627

```
# jdk_security_infra

security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java#emsigneccrootcag3 8328830 generic-all 
security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java#emsignrootcag1    8328830 generic-all
```

Comment (shipilev) [reply test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java:478]: Oooooh! So we found the CA that does not like GET requests, nice. I think these tests should be problemlisted with [JDK-8328830](https://bugs.openjdk.org/browse/JDK-8328830) subtask, and the fix for [JDK-8328830](https://bugs.openjdk.org/browse/JDK-8328830) would then fix/resolve the issue?

Comment (rhalade) [reply test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java:478]: Sure, I will do that then. Thanks!

**Review by rhalade — APPROVED** [review]

**Comment by shipilev** [comment]
CSR is done. I assume we are good here, and we can integrate?

**Comment by seanjmullan** [comment]
> CSR is done. I assume we are good here, and we can integrate?

Yes.

**Comment by shipilev** [comment]
All right, thanks! Here it goes.

/integrate

**Pull request closed without merging**